File: benchmarks/c/skbuff_alloc.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

typedef struct sk_buff {
    void *data;
    unsigned int len;
    unsigned int truesize;
} sk_buff;

sk_buff* alloc_skb(unsigned int size) {
    sk_buff *skb = malloc(sizeof(sk_buff));
    if (skb) {
        skb->data = malloc(size);
        skb->len = 0;
        skb->truesize = size;
    }
    return skb;
}

void free_skb(sk_buff *skb) {
    if (skb) {
        free(skb->data);
        free(skb);
    }
}

int main(int argc, char **argv) {
    int iterations = atoi(argv[1]);
    struct timespec start, end;

    clock_gettime(CLOCK_MONOTONIC, &start);

    for (int i = 0; i < iterations; i++) {
        sk_buff *skb = alloc_skb(1500);
        free_skb(skb);
    }

    clock_gettime(CLOCK_MONOTONIC, &end);

    double elapsed = (end.tv_sec - start.tv_sec) +
                     (end.tv_nsec - start.tv_nsec) / 1e9;
    printf("%.9f\n", elapsed);

    return 0;
}

File: benchmarks/rust/skbuff_alloc.rs
use std::alloc::{alloc, dealloc, Layout};
use std::env;
use std::time::Instant;

#[repr(C)]
struct SkBuff {
    data: *mut u8,
    len: u32,
    truesize: u32,
}

fn alloc_skb(size: usize) -> *mut SkBuff {
    unsafe {
        let layout = Layout::new::<SkBuff>();
        let skb = alloc(layout) as *mut SkBuff;
        if !skb.is_null() {
            let data_layout = Layout::from_size_align_unchecked(size, 8);
            let data = alloc(data_layout);
            (*skb).data = data;
            (*skb).len = 0;
            (*skb).truesize = size as u32;
        }
        skb
    }
}

fn free_skb(skb: *mut SkBuff) {
    unsafe {
        if !skb.is_null() {
            let data = (*skb).data;
            let size = (*skb).truesize as usize;
            let data_layout = Layout::from_size_align_unchecked(size, 8);
            dealloc(data, data_layout);

            let layout = Layout::new::<SkBuff>();
            dealloc(skb as *mut u8, layout);
        }
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let iterations: usize = args[1].parse().unwrap();

    let start = Instant::now();

    for _ in 0..iterations {
        let skb = alloc_skb(1500);
        free_skb(skb);
    }

    let elapsed = start.elapsed();
    println!("{:.9}", elapsed.as_secs_f64());
}
